File: kernel/rust/src/ipc/pipe.rs
//! Inter-process communication via pipes using ring buffers.
#![allow(dead_code)]

use alloc::boxed::Box;
use alloc::vec::Vec;
use core::cell::UnsafeCell;

pub const PIPE_BUFFER_SIZE: usize = 64 * 1024; // 64KB ring buffer

/// Ring buffer for pipe communication
pub struct RingBuffer {
    data: UnsafeCell<Vec<u8>>,
    read_pos: UnsafeCell<usize>,
    write_pos: UnsafeCell<usize>,
    size: usize,
}

impl RingBuffer {
    pub fn new() -> Self {
        let mut data = Vec::with_capacity(PIPE_BUFFER_SIZE);
        data.resize(PIPE_BUFFER_SIZE, 0);

        Self {
            data: UnsafeCell::new(data),
            read_pos: UnsafeCell::new(0),
            write_pos: UnsafeCell::new(0),
            size: PIPE_BUFFER_SIZE,
        }
    }

    /// Write data to the buffer. Returns number of bytes written.
    pub fn write(&self, buf: &[u8]) -> usize {
        unsafe {
            let data = &mut *self.data.get();
            let mut write_pos = *self.write_pos.get();
            let read_pos = *self.read_pos.get();

            let mut written = 0;

            for &byte in buf {
                // Check if buffer is full
                let next_write = (write_pos + 1) % self.size;
                if next_write == read_pos {
                    break; // Buffer full
                }

                data[write_pos] = byte;
                write_pos = next_write;
                written += 1;
            }

            *self.write_pos.get() = write_pos;
            written
        }
    }

    /// Read data from the buffer. Returns number of bytes read.
    pub fn read(&self, buf: &mut [u8]) -> usize {
        unsafe {
            let data = &*self.data.get();
            let mut current_pos = *self.read_pos.get();
            let write_pos = *self.write_pos.get();

            let mut read = 0;

            while read < buf.len() && current_pos != write_pos {
                buf[read] = data[current_pos];
                current_pos = (current_pos + 1) % self.size;
                read += 1;
            }

            *self.read_pos.get() = current_pos;
            read
        }
    }

    /// Check if buffer is empty
    pub fn is_empty(&self) -> bool {
        unsafe { *self.read_pos.get() == *self.write_pos.get() }
    }

    /// Check if buffer is full
    pub fn is_full(&self) -> bool {
        unsafe {
            let next_write = (*self.write_pos.get() + 1) % self.size;
            next_write == *self.read_pos.get()
        }
    }

    /// Get available space for writing
    pub fn available_space(&self) -> usize {
        unsafe {
            let read_pos = *self.read_pos.get();
            let write_pos = *self.write_pos.get();

            if write_pos >= read_pos {
                self.size - (write_pos - read_pos) - 1
            } else {
                read_pos - write_pos - 1
            }
        }
    }

    /// Get available data for reading
    pub fn available_data(&self) -> usize {
        unsafe {
            let read_pos = *self.read_pos.get();
            let write_pos = *self.write_pos.get();

            if write_pos >= read_pos {
                write_pos - read_pos
            } else {
                self.size - read_pos + write_pos
            }
        }
    }
}

unsafe impl Sync for RingBuffer {}

/// Pipe endpoints for IPC.
///
/// The read and write ends are **reference counted** (`readers`/`writers`),
/// not simple booleans: a pipe end may be held by several fds at once (e.g. the
/// shell holds a write end *and* hands a duplicate to a spawned child via
/// `spawn`). The end is only "closed" — and the peer only sees EOF / broken
/// pipe — when the last holder releases it. This is required for real pipelines
/// where an intermediate stage runs as a separate process.
pub struct Pipe {
    pub buffer: Box<RingBuffer>,
    pub blocked_reader: UnsafeCell<Option<u32>>, // TaskId waiting to read
    pub blocked_writer: UnsafeCell<Option<u32>>, // TaskId waiting to write
    readers: UnsafeCell<usize>,
    writers: UnsafeCell<usize>,
}

impl Pipe {
    pub fn new() -> Self {
        Self {
            buffer: Box::new(RingBuffer::new()),
            blocked_reader: UnsafeCell::new(None),
            blocked_writer: UnsafeCell::new(None),
            readers: UnsafeCell::new(0),
            writers: UnsafeCell::new(0),
        }
    }

    /// Acquire a reference to the read / write end (called when a `PipeSource` /
    /// `PipeSink` is created).
    pub fn add_reader(&self) {
        unsafe {
            *self.readers.get() += 1;
        }
    }
    pub fn add_writer(&self) {
        unsafe {
            *self.writers.get() += 1;
        }
    }

    /// Release one reference to the read end of the pipe.
    pub fn close_read(&self) {
        unsafe {
            let r = self.readers.get();
            *r = (*r).saturating_sub(1);
        }
    }

    /// Release one reference to the write end of the pipe.
    pub fn close_write(&self) {
        unsafe {
            let w = self.writers.get();
            *w = (*w).saturating_sub(1);
        }
    }

    /// True once every holder of the read end has released it.
    pub fn is_read_closed(&self) -> bool {
        unsafe { *self.readers.get() == 0 }
    }

    /// True once every holder of the write end has released it.
    pub fn is_write_closed(&self) -> bool {
        unsafe { *self.writers.get() == 0 }
    }

    /// Block a reader task
    pub fn block_reader(&self, task_id: u32) {
        unsafe {
            *self.blocked_reader.get() = Some(task_id);
        }
    }

    /// Block a writer task
    pub fn block_writer(&self, task_id: u32) {
        unsafe {
            *self.blocked_writer.get() = Some(task_id);
        }
    }

    /// Unblock reader and return the task id
    pub fn unblock_reader(&self) -> Option<u32> {
        unsafe {
            let task_id = *self.blocked_reader.get();
            *self.blocked_reader.get() = None;
            task_id
        }
    }

    /// Unblock writer and return the task id
    pub fn unblock_writer(&self) -> Option<u32> {
        unsafe {
            let task_id = *self.blocked_writer.get();
            *self.blocked_writer.get() = None;
            task_id
        }
    }
}

unsafe impl Sync for Pipe {}

File: memcontainers/kernel/zig/src/ipc/pipe.zig
//! src/ipc/pipe.zig — pipe ring buffers and reader/writer endpoint semantics (§2.6).
//!
//! Owns: stable ref-counted 64 KiB ring buffers with independent reader/writer ends, backpressure + wakeups, EOF on writer close, and EPIPE/signal on write without readers.
//! Invariants: stable pipe identity independent of scheduler storage moves (§2.6); correct fd-table interaction across spawn/exec/dup/close.
//! Oracle (behavior to match): kernel/rust/src/ipc/pipe.rs.
//! Not here: fd tables (task.zig); block/wake scheduling decisions (scheduler.zig). This file owns the buffer + endpoints; the scheduler owns WHEN a blocked task wakes.
//!
//! Scaffold status: implemented (Phase 4). `write`/`read` are the raw ring-buffer
//! primitives (byte count only, mirroring the oracle's `RingBuffer`); `isReadClosed`/
//! `isWriteClosed`/`isEmpty`/`isFull` are the predicates a future syscall layer composes
//! into EPIPE/SIGPIPE-on-write and EOF-on-read semantics — this file furnishes the state,
//! the decision to raise an errno or a signal belongs to whoever holds the owning `Task`
//! (out of scope here; see task.zig/scheduler.zig). The oracle's `blocked_reader`/
//! `blocked_writer` single-task-slot fields are NOT ported: they are dead code there too
//! (grep confirms no caller outside pipe.rs itself) — the real block/wake mechanism is
//! `BlockReason` + the scheduler's `blocked` map (task.zig/scheduler.zig).

const std = @import("std");

/// 64 KiB ring buffer capacity (matches the Rust oracle's `PIPE_BUFFER_SIZE`).
pub const PIPE_BUFFER_SIZE: usize = 64 * 1024;

/// A pipe: one fixed-capacity ring buffer plus reference-counted reader/writer ends.
///
/// Reference counts (not booleans) because several fds may share one end — e.g. a
/// pipeline stage spawned with an inherited write end. The peer only observes EOF (on
/// read) or a broken pipe (on write) once the LAST holder of the other end releases it
/// (`closeRead`/`closeWrite`). A freshly created pipe starts with zero readers AND zero
/// writers — whoever hands out each `Fd.pipe_read` / `Fd.pipe_write` must call
/// `addReader`/`addWriter` to register that handle, mirroring the oracle's
/// `PipeSource::new`/`PipeSink::new`.
pub const Pipe = struct {
    data: [PIPE_BUFFER_SIZE]u8 = undefined,
    read_pos: usize = 0,
    write_pos: usize = 0,
    readers: usize = 0,
    writers: usize = 0,

    /// Heap-allocate a pipe with a stable address (§2.6) — `BlockReason` and `Fd` carry
    /// raw `*Pipe` pointers into it, so it must never move for its lifetime.
    pub fn create(gpa: std.mem.Allocator) *Pipe {
        const self = gpa.create(Pipe) catch @panic("OOM");
        self.* = .{};
        return self;
    }

    /// Write as many bytes of `buf` as fit before the ring catches up to `read_pos`.
    /// Returns the count actually written — 0 iff the buffer is already full. Does NOT
    /// consult `isReadClosed` (see header: that decision lives with the caller).
    pub fn write(self: *Pipe, buf: []const u8) usize {
        var written: usize = 0;
        for (buf) |byte| {
            const next_write = (self.write_pos + 1) % self.data.len;
            if (next_write == self.read_pos) break; // full
            self.data[self.write_pos] = byte;
            self.write_pos = next_write;
            written += 1;
        }
        return written;
    }

    /// Read up to `buf.len` bytes. Returns the count actually read — 0 iff the buffer is
    /// empty. Does NOT consult `isWriteClosed` (see header: EOF-vs-would-block is the
    /// caller's decision).
    pub fn read(self: *Pipe, buf: []u8) usize {
        var n: usize = 0;
        while (n < buf.len and self.read_pos != self.write_pos) {
            buf[n] = self.data[self.read_pos];
            self.read_pos = (self.read_pos + 1) % self.data.len;
            n += 1;
        }
        return n;
    }

    /// Check if buffer is empty.
    pub fn isEmpty(self: *const Pipe) bool {
        return self.read_pos == self.write_pos;
    }

    /// Check if buffer is full.
    pub fn isFull(self: *const Pipe) bool {
        return (self.write_pos + 1) % self.data.len == self.read_pos;
    }

    /// Register a new holder of the read / write end — a fresh `Fd.pipe_read`/
    /// `Fd.pipe_write` pointing here, including a `dup`'d or child-inherited copy.
    pub fn addReader(self: *Pipe) void {
        self.readers += 1;
    }
    pub fn addWriter(self: *Pipe) void {
        self.writers += 1;
    }

    /// Release one holder's claim on the read / write end. Idempotent (saturating) —
    /// safe even if called more times than `addReader`/`addWriter` ever were.
    pub fn closeRead(self: *Pipe) void {
        self.readers -|= 1;
    }
    pub fn closeWrite(self: *Pipe) void {
        self.writers -|= 1;
    }

    /// True once every holder of the read end has released it — a writer's cue to fail
    /// the write (EPIPE) / raise SIGPIPE instead of blocking forever.
    pub fn isReadClosed(self: *const Pipe) bool {
        return self.readers == 0;
    }

    /// True once every holder of the write end has released it — a reader's cue that an
    /// empty buffer now means EOF, not "block and retry".
    pub fn isWriteClosed(self: *const Pipe) bool {
        return self.writers == 0;
    }
};
